File: LEETCODE/LEETCODE-CPP/543.cpp
//543. Diameter of Binary Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res=0;
        height(root,res);
        return res;
    }
    
     int height(TreeNode* root,int &res){
        if(root){
            int left=height(root->left,res);
            int right=height(root->right,res);
            
            res=max(res,left+right);
            
            return max(left,right)+1;
        }
        else
             return 0;
    }
};

File: LEETCODE/LEETCODE-PYTHON/543.py
# 543. Diameter of Binary Tree
# https://leetcode.com/problems/diameter-of-binary-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter=0
        def helper(root):
            nonlocal diameter
            if root is None:
                return 0

            l= helper(root.left)
            r=  helper(root.right)
            diameter=max(diameter, l+r)

            return  max(l,r)+1

        helper(root)
        return diameter

